packages/shared/pkg/utils/warmpool.go

```diff
@@ -0,0 +1,321 @@
+	}
+	recordFailure := func(reason string, attrs ...attribute.KeyValue) {
+		attrs = append(attrs, failureAttr(), reasonAttr(reason))
+		wp.getCounter.Add(ctx, 1, metric.WithAttributes(attrs...))
+	}
+
+	// early check to bail
+	if err := wp.isClosed(ctx); err != nil {
+		recordFailure("closed")
+
+		return item, err
+	}
+
+	// get from the reusable pool first
+	select {
+	case <-ctx.Done():
+		recordFailure("canceled")
+
+		return item, ctx.Err()
+	case <-wp.done:
+		recordFailure("closed")
+
+		return item, ErrClosed
+	case s := <-wp.reusableItems:
+		telemetry.ReportEvent(ctx, fmt.Sprintf("reused %s", wp.name))
+		recordSuccess("used")
+
+		return s, nil
+	default:
+	}
+
+	// if that didn't work, get from whichever pool gets an item first
+	select {
+	case <-ctx.Done():
+		recordFailure("canceled")
+
+		return item, ctx.Err()
+	case <-wp.done:
+		recordFailure("closed")
+
+		return item, ErrClosed
+	case s := <-wp.reusableItems:
+		telemetry.ReportEvent(ctx, fmt.Sprintf("reused %s", wp.name))
+		recordSuccess("used")
+
+		return s, nil
+	case s := <-wp.freshItems:
+		telemetry.ReportEvent(ctx, fmt.Sprintf("new %s", wp.name))
+		recordSuccess("fresh")
+
+		return s, nil
+	}
+}
+
+// Return an item to the pool.
+// - puts the item in `reusableItems` to get priority when being returned.
+// - destroy the item and return if the pool has already been closed.
+// - if the item fails to be pushed back into the channel for any reason, destroy it.
+func (wp *WarmPool[T]) Return(ctx context.Context, item T) {
+	wp.wg.Go(func() {
+		recordSuccess := func(attrs ...attribute.KeyValue) {
+			attrs = append(attrs, successAttr())
+			wp.returnedCounter.Add(ctx, 1, metric.WithAttributes(attrs...))
+		}
+
+		recordFailure := func(reason string, attrs ...attribute.KeyValue) {
+			attrs = append(attrs, failureAttr(), reasonAttr(reason))
+			wp.returnedCounter.Add(ctx, 1, metric.WithAttributes(attrs...))
+		}
+
+		if err := wp.isClosed(ctx); err != nil {
+			recordFailure("closed")
+
+			return
+		}
+
+		// try to push item in to reusable pool first
+		select {
+		case wp.reusableItems <- item:
+			telemetry.ReportEvent(ctx, fmt.Sprintf("returned %s", wp.name))
+			recordSuccess()
+
+			return
+		default:
+		}
+
+		// if that didn't work, keep trying to push, but listen for failures as well
+		select {
+		case wp.reusableItems <- item:
+			telemetry.ReportEvent(ctx, fmt.Sprintf("returned %s", wp.name))
+			recordSuccess()
+
+		case <-ctx.Done():
+			wp.beginDestroy(ctx, item, "return failed due to canceled context")
+			recordFailure("canceled")
+
+		case <-wp.done:
+			wp.beginDestroy(ctx, item, "return failed due to closed pool")
+			recordFailure("closed")
+		}
+	})
+}
+
+// Close destroys all items and prevents the pool from creating/returning anymore.
+func (wp *WarmPool[T]) Close(ctx context.Context) error {
+	logger.L().Info(ctx, "Closing pool")
+
+	if err := wp.isClosed(ctx); err != nil {
+		return err
+	}
+
+	wp.doneOnce.Do(func() {
```

Comment:
nit: why do you close freshItems in different function? Why not to keep it consistent?

Reply:
Good point, the mechanism in Populate is much cleaner and easier to get right